### DISCUSSION TITLE
Parsing XPath predicates

### DISCUSSION THREAD

**Discussion by rengit62** [OP-TED/eForms-SDK#154] [Q&A] [answered]
Hi all !
what about field BT-01(c) ?
"fields" : [ {
    "id" : "BT-01(c)-Procedure",
**************************
    "xpathAbsolute" : "/*/cac:TenderingTerms/cac:ProcurementLegislationDocumentReference/cbc:ID[not(text()=('CrossBorderLaw','LocalLegalBasis'))]",
    "xpathRelative" : "cac:ProcurementLegislationDocumentReference/cbc:ID[not(text()=('CrossBorderLaw','LocalLegalBasis'))]",

I guess .. if id is not equal to CrossBorderLaw' or LocalLegalBasis then BT-01(c)-Procedure
else BT-09(a)-Procedure ( if CrossBorderLaw)
does it ?
And if it does, is there an EFX api to parse the predicate ? 
Thanks !
Renato

I tried with

**Comment by YvesJo** [comment, marked as answer]
Hi,
The statement is correct:

> if ID is not equal to CrossBorderLaw' or LocalLegalBasis then BT-01(c)-Procedure
> else BT-09(a)-Procedure ( if CrossBorderLaw)

Of course for BT-01(c)-Procedure, ID shall exist.
Also to complete the statement, when no ID is available and not about CrossBorderLaw, then BT-01(e)-Procedure

**Reply by rengit62** [reply]
Thanks Yves
and for the second question ?
How can I parse the predicate ? Is there an EFX api ?
Bye
Renato

**Comment by rousso** [comment]
Hi @rengit62,

I think there is a misunderstanding in your question.

You are asking about parsing XPath predicates using an EFX parser. 
EFX has nothing to do with XPath. It is a different language. The EFX parser only parses EFX.

If you want to parse the XPath predicate you need an XPath parser.

You can create an XPath parser using ANLR4. But there are probably many other ways to do it.
What we do to parse XPath in the EFX-Toolkit is use an ANTLR4 based XPath parser.

The XPath grammar is included in the EFX-Toolkit's source code. The original source of the XPath grammar we used is here: https://github.com/antlr/grammars-v4/tree/master/xpath. We use XPath 2.0

I have a question however: Why do you want to parse the XPath predicate? 
What are you going to do with it?

Le me know.
Regards,
Yannis

**Reply by rengit62** [reply]
Ok Yannis, thanks
I try to explain, let get notice subtype 1 :
I did a java main to build the model and another java main to build the xml notice, by the model, the xsd's and some mocked values.
When the second main 'meets' an element like : 
BT-05(a)-notice											cbc:IssueDate
it makes an xml tag  <<cbc:IssueTime> 'son of root'  ( <PriorInformationNotice)

<PriorInformationNotice xmlns="urn:oasis:names:specification:ubl:schema:xsd:PriorInformationNotice-2" *************>
***********************************************	
	<cbc:IssueDate>2022-11-28+10:35</cbc:IssueDate>
	
and so on 
what should this java main do when from the model meets an element like : 
BT-01(c)-Procedure											cac:ProcurementLegislationDocumentReference/cbc:ID[not(text()=('CrossBorderLaw','LocalLegalBasis'))]
? 
Usually xpath is used to make query on a xml .
I guess here it is used to express different xml tags depending on some values, as written in doc
https://docs.ted.europa.eu/eforms/1.4.0/schema/procedure-lot-part-information.html#
But the value are set using BT* identifier.
So what do you suggest to manage xpath predicate in this scenario ?
Regards and thanks,
Renato
